:zap: Added native construction of two-target controlled-gate DDs

BertiFlorea opened the pull request:
## Description
The purpose of this PR is to add the extra logic for two-target controlled-gate DDs mentioned in https://github.com/cda-tum/mqt-core/issues/334#issue-1758497551. The problem was that the current implementation of the `makeTwoQubitGateDD` only worked for non-controlled two-target gates. Additionally, what we also want to achieve with this PR is to get rid of all the extra methods using the previous workaround (i.e.https://github.com/cda-tum/dd_package/blob/7c9644a2a65742bb580e3e5fefeac075c9534351/include/dd/Package.hpp#L629-L810).

At the moment the updated version of `makeTwoQubitGateDD` is passing the CI, but there are still some TODOs left:
- [x] test for equivalence between old-fashioned workaround methods and `makeTwoQubitGateDD`
- [x] update the code across the package (e.g. getStandardOperationDD)
- [x] verify the functionality by testing for edge cases

<!---Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.-->

Fixes #334 <!--- Replace (issue) with the issue number that is fixed by this pull request. -->

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

burgholzer reviewed (changes requested):
Overall the changes here are looking good and it seems like this should be working.
The biggest thing that is missing is tests of the new functionality. That's also why the code coverage check is failing. Almost none of the new lines (part of the patch) are covered by tests.

This should be quickly resolved though once the existing workaround methods are removed.
I would propose to turn the existing workarounds for constructing controlled two-target DDs into tests for the new functionality. That way, the existing code is not lost and is useful for testing.

BertiFlorea commented:
At the moment, the following progress has been made:
1. The `makeTwoQubitGateDD` constructs the DD bottom-up correctly. 🎉 
2. I have updated the `makePeresDD` and `makePeresInvDD` workaround methods because they provided the inverse output instead of the desired one. (i.e. makePeresDD was producing the inverse DD of the Peres gate)
3. I have updated the code across the code base to rely on the native construction of two-target controlled-gate DDs instead of using the old workaround methods. Additionally, I used them to verify the functionality of our method by comparing the outputs (e.g. compared the output generated by `makeTwoQubitGateDD` for a swap matrix with `makeSWAPDD`).
Note: Here there were some already existing methods which were checking this, however, there was only one missing test case for the Peres gate.
4. To improve the code coverage I have added one test to verify all the edge cases that were not covered by the existing tests. (Update after the CI has finished: the test still needs some improvement, I'll come back with an update for this one)

Apart from point 4 (which is on my TODO list), the only question left from my side is if we still intend to unify `makeTwoQubitGateDD` and `makeGateDD` since it was mentioned in #334.

burgholzer commented on include/dd/Operations.hpp:
```diff
@@ -107,137 +107,57 @@ getStandardOperationDD(const qc::StandardOperation* op, Package<Config>& dd,
+    break;
   case qc::iSWAP:
-    if (inverse) {
-      return dd.makeiSWAPinvDD(nqubits, controls, target0, target1, startQubit);
-    }
-    return dd.makeiSWAPDD(nqubits, controls, target0, target1, startQubit);
```
The corresponding routines in the `Package.hpp` file can all be deleted now. That's actually what I meant with one of my earlier comments that all these functions for creating controlled two-target functions can be extracted into tests.
Just remove all that code from the main package (`Package.hpp`) and move it to the tests. You can just copy (and slighlty adjust) the definitions from the `Package.hpp` file and move them to the `test_dd_functionality.cpp` test file.
This should reduce the number of functions in the main package file considerably.

BertiFlorea replied:
Done 👍🏼

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -64,7 +64,7 @@ TEST_P(DDFunctionality, standardOpBuildInverseBuild) {
   auto gate = static_cast<qc::OpType>(GetParam());
 
-  qc::StandardOperation op;
+  qc::StandardOperation op, op_neg;
```
First of all, you can see some of the clang-tidy warnings here in the "Files Changed" tab.
These should be easy to address.
In this case here, you do not seem to be using `op_neg` for the test at all, so it could be removed entirely.

BertiFlorea replied:
Done 👍🏼

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -121,7 +121,7 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
   auto gate = static_cast<qc::OpType>(GetParam());
 
-  qc::StandardOperation op;
+  qc::StandardOperation op, op_neg;
```
I'd slightly prefer to refactor this into two separate tests. Let's keep the original test and extract the negative control one into a separate function.
Then, you can also directly test the single-target gates with negative controls, which should yield even better coverage.

BertiFlorea replied:
Done 👍🏼

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -172,6 +177,41 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
+  }
+}
+
+TEST_F(DDFunctionality, twoTargetControlledGateDDEdgeCases) {
+  using namespace qc::literals;
+  auto newNQubits = 12U;
+  auto newDD = std::make_unique<dd::Package<>>(newNQubits);
+  auto gate = SWAP;
+
+  qc::StandardOperation op, op_neg;
+  op = qc::StandardOperation(newNQubits, Controls{1, 5, 7, 11}, 3, 9, gate);
+  op_neg = qc::StandardOperation(newNQubits, Controls{1_nc, 5_nc, 7_nc, 11_nc},
+                                 3, 9, gate);
+
+  ASSERT_NO_THROW(
+      { e = newDD->multiply(getDD(&op, *newDD), getInverseDD(&op, *newDD)); });
+  newDD->incRef(e);
+
+  ASSERT_NO_THROW({
+    d = newDD->multiply(getDD(&op_neg, *newDD), getInverseDD(&op_neg, *newDD));
+  });
+  newDD->incRef(d);
+
+  // the DD of the identity needs to be reconstructed as well
+  ident = newDD->makeIdent(newNQubits);
+
+  EXPECT_EQ(ident, e);
+  EXPECT_EQ(ident, d);
```
This test seems oddly specific to me. What part of that is not already covered by existing tests?
If it's the part about the negative controls, then this should be fairly easy to incorporate into the existing tests that test out all the possibilities of combining controls and targets by adding another for loop that influences the type of control.

burgholzer reviewed (changes requested):
Overall this is looking really good already. Many Thanks 🎉 
The Python failures should be fixed after #536 is in. Then you can just merge the main branch in or rebase your branch on main.
There still are some warnings left to be fixed and I also added some comments that would be good to address before this gets merged.
Most importantly, the main library code in `Package.hpp` can be significantly reduced by moving the respective special purpose construction methods to the tests. This should slim down the library quite a bit.

As for your question on combining the two gate DD creation methods: I do not think that is necessary as part of this PR. Content-wise, this PR is great as it is and definitly enough to close the respective issue 👍🏼

BertiFlorea commented:
I have just added all the requested changes. The code coverage failed but according to the error message, it seems that it might be a connection problem (I might be wrong still).

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -174,6 +174,63 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
+  qc::StandardOperation op;
+  switch (gate) {
+  case qc::GPhase:
+    op = qc::StandardOperation(nqubits, Controls{0}, Targets{}, gate,
```
```suggestion
    op = qc::StandardOperation(nqubits, Controls{0_nc}, Targets{}, gate,
```

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -174,6 +174,63 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
+                               std::vector{dist(mt)});
+    break;
+  case qc::U:
+    op = qc::StandardOperation(nqubits, 0, 1, gate,
```
```suggestion
    op = qc::StandardOperation(nqubits, Controls{0_nc}, 1, gate,
```

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -174,6 +174,63 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
+                               std::vector{dist(mt), dist(mt), dist(mt)});
+    break;
+  case qc::U2:
+    op = qc::StandardOperation(nqubits, 0, 1, gate,
```
```suggestion
    op = qc::StandardOperation(nqubits, Controls{0_nc}, 1, gate,
```

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -174,6 +174,63 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
+  case qc::RY:
+  case qc::RZ:
+  case qc::P:
+    op = qc::StandardOperation(nqubits, 0, 1, gate, std::vector{dist(mt)});
```
```suggestion
    op = qc::StandardOperation(nqubits, Controls{0_nc}, 1, gate, std::vector{dist(mt)});
```

burgholzer commented on test/dd/test_dd_functionality.cpp:
```diff
@@ -174,6 +174,63 @@ TEST_P(DDFunctionality, controlledStandardOpBuildInverseBuild) {
+                               std::vector{dist(mt), dist(mt)});
+    break;
+  default:
+    op = qc::StandardOperation(nqubits, 0, 1, gate);
```
```suggestion
    op = qc::StandardOperation(nqubits, Controls{0_nc}, 1, gate);
```

burgholzer reviewed (approved):
Great. Many thanks for another nice contribution! 🎉 
I'll just apply some finishing touches and then proceed to merge 🔀
